ftw/zipextract/implementations_at.py

```diff
@@ -0,0 +1,19 @@
+from ftw.zipextract.implementations_base import ObjectCreatorBase
+from ftw.zipextract.interfaces import IFileCreator
+from ftw.zipextract.interfaces import IFolderCreator
+from zope.component import adapts
+from zope.interface import implements
+from Products.ATContentTypes.content.file import IATFile
+from Products.ATContentTypes.content.folder import IATFolder
+
+
+class ATFileCreator(ObjectCreatorBase):
+    implements(IFileCreator)
+    adapts(IATFile)
```

Comment:
How is this adapter adapting an IATFile? It is a factory adapter, so it will create the file. At the point the adapter is created it cannot yet exist.

Reply:
Yes I know this is weird and I'd be happy about a better solution. For the `FolderCreator`s the adapter takes the parent folder to know which type of folder we are working with, and hence what `portal_type` to create. For the `FileCreator`s I use the current zip file (i.e. the context of the zipextracter) and adapt that. This seemed strange to me when I wrote it, but I couldn't think of another way.

Reply:
Ah, now I understand it.

I propose to change it so that the adapted context is always the future parent. This would work with our use cases. The current implementation also does not support adding DX-objects within AT-objects or vice-versa.

For example in workspaces:
Files and folders can both be added within `Workspace`-objects or within `TabbedViewFolder`-objects. Therefore I'd let both adapt `Products.CMFCore.interfaces.IFolderish`. `IFolderish` is provided by all folderish Plone objects (AT and DX).

The problem with this change is that we then cannot distinguish AT and DX creator adapters, because we are adapting the parent. We also cannot adapt the future object since it is not yet created at this point.
I would solve this problem by also adapting the FTI (Factory Type Information) for the portal-type to be created (get it with `getToolByName(context, 'portal_types').get('File')`.

That means that we need to move the decision, which portal type to create, outside of this creator. We also may need to support creating multiple file types depending on the content. For example in standard Plone we could create `Image`-objects for images. When `ftw.mail` is installed we could create a mail-object for `.eml`-files.

So the adapters could look like that:
```python
@implementer(IFolderCreator)
@adapter(Interface, Interface, IDexterityFTI)
class DXFolderCreator(object):

    def __init__(self, container, request, fti):
        self.container = container
        self.request = request
        self.fti = fti

    def create(self, title):
        return createContentInContainer(self.container, self.fti.factory, title=title)


@implementer(IFileCreator)
@adapter(Interface, Interface, IDexterityFTI)
class DXFileCreator(object):

    def __init__(self, container, request, fti):
        self.container = container
        self.request = request
        self.fti = fti

    def create(self, filename, blobfile):
        obj = createContentInContainer(self.container, self.fti.factory, title=filename)
        field = IPrimaryFieldInfo(obj).field
        field.set(obj, NamedBlobFile(obj))
        return obj

    
@implementer(IFolderCreator)
@adapter(Interface, Interface, Interface)
class ATFolderCreator(object):

    def __init__(self, container, request, fti):
        self.container = container
        self.request = request
        self.fti = fti

    def create(self, title):
        normalized_id = normalize_id(name)
        chooser = INameChooser(context)
        new_id = chooser.chooseName(normalized_id, context)
        return self.container.invokeFactory(self.fti.factory, new_id, title=title)

    
@implementer(IFileCreator)
@adapter(Interface, Interface, Interface)
class ATFolderCreator(object):

    def __init__(self, container, request, fti):

        self.request = request
        self.fti = fti

    def create(self, filename, blobfile):
        normalized_id = normalize_id(filename)
        chooser = INameChooser(context)
        new_id = chooser.chooseName(normalized_id, context)
        obj = self.container.invokeFactory(self.fti.factory, new_id, title=title)
        obj.getPrimaryField().set(obj, blobfile)
        return obj```

Thoughts and hints:
- I combined creating and setting the blob file, because I find this better understandable.
- The DX-FTI provides IDexterityFTI, but the AT-FTI does not provide interfaces which are not provided by DX, that's why the AT adapters adapt "Interface" at this position.
- I also mixed the request in: this allows for easier customization with browser layers in the future.
- I'm using the primary field instead of hard-coded "file" field name; the primary field is usually the correct one.

When using the adapter you could first query for a named adapter, then fallback to the default
adapter. This will also allow us to e.g. customizate the "Mail" creator but not the "File" creator.

```python
def get_creator(interface, container, portal_type):
    fti = getToolByName(container, 'portal_types').get('portal_type')
    creator = queryMultiAdapter((container, container.REQUEST, fti), interface, name=portal_type)
    if creator is not None:
        return creator
    else:
        return getMultiAdapter((container, container.REQUEST, fti), interface)


get_creator(IFolderCreator, context, 'TabbedViewFolder')
get_creator(IFileCreator, context, 'File')
```

Reply:
Hum, I think we'll have to discuss that tomorrow. I modified the interfaces, but it does not really change the issue knowing what portal type will be created. Or maybe there is something I don't understand.
`IFolderCreator` and `IFileCreator` get invoked from `zipextracter`. And there I still only have the container where the new item will be created and a filename or foldername. So I still end up using the container's portal_type as argument of `get_creator`.

Also I was not able to use the `fti` for archetypes. `invokeFactory` expects a `portal_type` as first argument and `fti.portal_type` is always `Plone Site` no matter what the container was. And I don't know what to do with `ftw.factory` in that case. The way you wrote it `obj = self.container.invokeFactory(self.fti.factory, new_id, title=title)` will not work (as said it expects a `portal_type`).